feat: table grouping and subtotals

agha4to opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

Closes: https://github.com/lightdash/lightdash/issues/3120

### Description:
<!-- Add a description of the changes proposed in the pull request. -->

Adds a "Show subtotals" checkbox to chart configuration which groups rows based on dimension values, and provides subtotals for sum and count metrics.

<!-- Even better add a screenshot / gif / loom -->

![image](https://github.com/lightdash/lightdash/assets/138995798/891e2c7e-8e05-4ac3-8069-54c7368f9648)

![image](https://github.com/lightdash/lightdash/assets/138995798/a3dc9151-f3f9-43f6-badc-5cb9da2ff1bf)


### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

TuringLovesDeathMetal reviewed (changes requested):
This is amazing! Well done 🚀 

I've been playing around with it in this PR environment here: 

Looks, and feels really awesome and super excited to share this with people. 

Just three changes from me (sorry, I realize I missed some of these on the acceptance criteria, so this is mostly up to me missing details in the spec!):

- Interactivity on subtotalled cells should be turned off (`copy value`, `show underlying data`, etc. all seem to "pull" from the first value in the subtotal, so I think let's just turn it off for this first version, then we can always add it in later when people ask for it!). The `copy` seems to be working for the `toggled`/`grouped` cell, so we can leave that in if it's easy 👀 

<img width="1094" alt="image" src="https://github.com/lightdash/lightdash/assets/31848148/74461e35-dee7-49d3-a9e8-236a6e26bf9f">

- Can we make the number shown in the toggle represent the total number of underlying rows (instead of the number of rows in the _**next**_ grouping level? E.g. the `date month` here, I would expect this to say `19` instead of `4`. 

<img width="1027" alt="image" src="https://github.com/lightdash/lightdash/assets/31848148/b14951d1-917c-4cf7-9961-89fab08971bf">

- Can we make it so that `show subtotals` only appears when there are 2 or more dimensions in the results table? (e.g. here, the `show subtotals` option shouldn't appear)

<img width="1443" alt="image" src="https://github.com/lightdash/lightdash/assets/31848148/4dc915d4-10b0-4054-a825-f0b3f002d024">

IrakliJani commented on packages/frontend/src/components/VisualizationConfigs/TableConfigPanel/GeneralSettings.tsx:
```diff
@@ -248,6 +250,25 @@ const GeneralSettings: FC = () => {
                     }}
                 />
+                <Tooltip
+                    disabled={!canUsePivotTable}
```
I think the variable name is misleading. We should either introduce another variable that better describes whether the pivot table is enabled or just rename this one to `isPivotTableEnabled` or something similar.

IrakliJani commented on packages/frontend/src/components/common/Table/ScrollableTable/TableBody.tsx:
```diff
@@ -79,34 +81,89 @@ const TableRow: FC<TableRowProps> = ({
-                            cell.getContext(),
+                        {cell.getIsGrouped() ? (
+                            <>
+                                <button
```
this can be a mantine's `ActionIcon` component with minimal styling (https://v6.mantine.dev/core/action-icon/)

agha4to replied:
Tried ActionIcon, but was not rendering nicely with both icon and count label being clickable. Tried UnstyledButton, but settled on Button with variant 'subtle`, but please let me know if we should change it.

IrakliJani commented on packages/frontend/src/components/common/Table/ScrollableTable/TableBody.tsx:
```diff
@@ -79,34 +81,89 @@ const TableRow: FC<TableRowProps> = ({
+                        {cell.getIsGrouped() ? (
+                            <>
+                                <button
+                                    {...{
+                                        onClick: (e) => {
+                                            e.stopPropagation();
+                                            e.preventDefault();
+                                            toggleExpander();
+                                        },
+                                        style: {
+                                            cursor: row.getCanExpand()
+                                                ? 'pointer'
+                                                : 'normal',
+                                            border: 'none',
+                                            background: 'none',
+                                            float: 'left',
+                                        },
+                                    }}
```
I think an extra spread here is not a requirement

agha4to replied:
Nice catch!

IrakliJani commented on packages/frontend/src/components/common/Table/ScrollableTable/TableBody.tsx:
```diff
@@ -79,34 +81,89 @@ const TableRow: FC<TableRowProps> = ({
+                            flexRender(
+                                cell.column.columnDef.cell,
+                                cell.getContext(),
+                            )
```
Looking at this file, I feel like there's a possibility to clean up some of the styling code and leverage Mantine components more. But we can do that at the end of the stage.

IrakliJani replied:
this part is very clear. thanks @agha4to !

IrakliJani commented on packages/frontend/src/components/common/Table/ScrollableTable/TableHeader.tsx:
```diff
@@ -15,12 +16,42 @@ import { HeaderDndContext, HeaderDroppable } from './HeaderDnD';
+
+    useEffect(() => {
+        if (showSubtotals) {
+            const groupedColumns: any = columns
+                .filter(
+                    (col: any) =>
+                        col.meta?.item?.fieldType === FieldType.DIMENSION,
+                )
+                .map((col: any) => col.id);
+            const sortedColumns: any = table
+                .getState()
+                .columnOrder.reduce((acc: any[], sortedId) => {
```
This code makes sense, however, it would be nice if we could avoid having `ANY` as it often introduces unnecessary runtime bugs. I'm happy to chime in here and update types.

agha4to replied:
I got rid of the `ANY`'s but not sure my code to replace them is what you have in mind. Please let me know.

IrakliJani commented on packages/frontend/src/components/common/Table/TableProvider.tsx:
```diff
@@ -75,10 +80,12 @@ const calculateColumnVisibility = (columns: Props['columns']) =>
     ...rest
 }) => {
     const { data, columns, columnOrder, pagination } = rest;
+    const [grouping, setGrouping] = React.useState<GroupingState>([]);
```
```suggestion
    const [grouping, setGrouping] = useState<GroupingState>([]);
```

IrakliJani commented on packages/frontend/src/hooks/tableVisualization/getDataAndColumns.tsx:
```diff
@@ -55,6 +66,32 @@ const getDataAndColumns = ({
             }
             const headerOverride = getFieldLabelOverride(itemId);
 
+            const anyItem = item as any;
```
same about `any` type in this function

IrakliJani commented:
Hey @agha4to! This looks really good! I did the first round of code review and added comments, mostly about minor details. Let me know if you have any questions or need any help. I'll do a second round of review after you make the changes that were mentioned above 🙌

TuringLovesDeathMetal commented:
Nice! looks great :) 

There's just one tiny thing - it looks like we removed the items from the interactivity menu when you click on a subtotal, but there's still a little box that appears. Is there any way we can remove that? 

<img width="391" alt="image" src="https://github.com/lightdash/lightdash/assets/31848148/0b5b311a-09b6-4e43-9f8b-2a91cf54d4bf">

agha4to commented:
Committed changes for the issues you found:
0) Any non-number in a column now stops subtotal loop and displays blank cell instead of subtotal value
1) Cell dropdown menu artifact removed for empty group cells and subtotal cells
2) Subtotals checkbox shows as unchecked when disabled

TuringLovesDeathMetal reviewed (approved):
wahoooo! looks great :)

IrakliJani commented on packages/frontend/src/components/common/Table/getGroupedRowModelLightdash.ts:
```diff
@@ -0,0 +1,225 @@
+// https://tanstack.com/table/v8/docs/guide/row-models#customizefork-row-models
+
+function groupBy<TData extends RowData>(rows: Row<TData>[], columnId: string) {
+    const groupMap = new Map<any, Row<TData>[]>();
+
+    return rows.reduce((map, row) => {
+        // This line changed from original since getGroupingValue does not exist here as it should.
+        // const resKey = `${row.getGroupingValue(columnId)}`
+        const resKey = (row.original as any)[columnId].value.raw;
```
can these `any` be `unknown` (without a lot of changes)?

agha4to replied:
Yes, looks OK.

IrakliJani replied:
perfect, thanks!

IrakliJani commented on packages/frontend/src/components/common/Table/getGroupedRowModelLightdash.ts:
```diff
@@ -0,0 +1,225 @@
+
+                const groupedFlatRows: Row<TData>[] = [];
+                const groupedRowsById: Record<string, Row<TData>> = {};
+                // const onlyGroupedFlatRows: Row[] = [];
+                // const onlyGroupedRowsById: Record<RowId, Row> = {};
+                // const nonGroupedFlatRows: Row[] = [];
+                // const nonGroupedRowsById: Record<RowId, Row> = {};
```
can we remove commented out code from this file? or maybe we could add a comment for the context why it's commented out?

agha4to replied:
This file is a copy of https://github.com/TanStack/table/blob/main/packages/table-core/src/utils/getGroupedRowModel.ts which has these lines commented out. We can remove the lines though. The only reason to keep them is to make comparison to the source file more straightforward in the future.

IrakliJani replied:
fair enough, let's keep them! thanks Tony!

agha4to replied:
Actually, I think it is no big deal to remove them. It does not make the comparison to the original especially difficult.

IrakliJani commented on packages/frontend/src/hooks/tableVisualization/useTableConfig.ts:
```diff
@@ -158,13 +162,26 @@ const useTableConfig = (
+        });
+    }, [columnOrder, itemsMap]);
+
+    const canUseSubtotals = dimensions.length > 1;
```
can we use `resultsData && resultsData.metricQuery.dimensions.length > 0` instead? if so we could avoid `useMemo` above

agha4to replied:
`resultsData.metricQuery.dimensions` does not update when dimensions are removed, so if all but one dimension is removed and "Configure" is chosen then the "Show subtotals" checkbox will still be enabled because `resultsData.metricQuery.dimensions` still has the original number of dimensions while the `dimensions` constant calculated here reflects the change to one dimension.

IrakliJani replied:
that makes sense! `useTableConfig` hook is so confusing!

IrakliJani commented on packages/frontend/src/hooks/tableVisualization/getDataAndColumns.tsx:
```diff
@@ -100,6 +165,24 @@ const getDataAndColumns = ({
                         isVisible: isColumnVisible(itemId),
                         frozen: isColumnFrozen(itemId),
                     },
+
+                    // Some features work in the TanStack Table demos but not here, for unknown reasons.
+                    // For example, setting grouping value here does not work. The workaround is to use
+                    // a custom getGroupedRowModel.
+                    // getGroupingValue: (row) => { // Never gets called.
+                    //     const value = row[itemId]?.value.raw;
+                    //     return value === null || value === undefined ? 'null' : value;
+                    // },
+                    // aggregationFn: 'sum', // Not working.
+                    // aggregationFn: 'max', // At least results in a cell value, although it's incorrect.
+
```
thanks, this is super helpful!

IrakliJani reviewed (commented):
Hey @agha4to, I did a thorough review of the code. I feel like we've addressed all aspects of the feature and the bugs as well. This is looking great; I just added a couple of questions/minor suggestions.

Let me know if you have any questions!

agha4to commented on packages/frontend/src/components/SimpleTable/index.tsx:
```diff
@@ -139,6 +141,13 @@ const SimpleTable: FC<SimpleTableProps> = ({
                         );
+                    const cell = props.cell;
+                    if (
+                        cell.getValue() === undefined ||
```
That's a really nice catch. It's especially good because the changes in this file are no longer needed, since the changes to TableBody.tsx were recently added. Removes redundant code.

IrakliJani replied:
perfect!

agha4to commented on packages/frontend/src/components/common/Table/ScrollableTable/TableBody.tsx:
```diff
@@ -79,34 +92,88 @@ const TableRow: FC<TableRowProps> = ({
+                        : undefined;
+
+                const suppressContextMenu =
+                    cell.getValue() === undefined ||
+                    cell.getIsPlaceholder() ||
+                    cell.getIsAggregated();
```
`getValue()` was needed when this code ran (above) in `SimpleTable/index.tsx` but not here, perhaps due to the recent change where any non-Number value in a column results in subtotal short-circuiting.

agha4to commented on packages/frontend/src/components/common/Table/ScrollableTable/TableHeader.tsx:
```diff
@@ -15,12 +16,44 @@ import { HeaderDndContext, HeaderDroppable } from './HeaderDnD';
+                    if (groupedColumns.includes(sortedId)) acc.push(sortedId);
+                    return acc;
+                }, []);
+            sortedColumns.pop();
```
The last dimension column essentially groups rows for each unique value in that column. Without the `pop()` each row would have an extra row group containing 1 item.

IrakliJani replied:
gotcha, thanks!

agha4to commented:
@TuringLovesDeathMetal, should the "Show subtotals" checkbox also appear unchecked when there are fewer than 2 dimensions?
<img width="310" alt="image" src="https://github.com/lightdash/lightdash/assets/138995798/fcb8ad52-a55f-48a6-b874-a20d5810b1a1">

TuringLovesDeathMetal commented:
@agha4to - oooh, ya, that would be nice! 

It's not a blocking change from me though - so if you feel like doing it you can :)

IrakliJani commented:
@agha4to, thank you for your comment replies and for clarifying my questions. It was super helpful.

I've pushed a few minor updates to your branch:
- Slight style updates to the group collapse button.
- Ran the API generation script for the backend.

---

I noticed rather odd behavior in the minimal pages' table when collapsing/expanding the groups.

"explore page" and "dashboard" seem fine. The bug is only present on "minimal explore page" (which is not supposed to be interactive). However, This raised a valid concern about the nature of the bug and whether it could affect anything else:


https://github.com/lightdash/lightdash/assets/962095/95a07d7d-916c-464b-a198-dc9eaa4c5521

---


I propose that we deploy this behind a feature flag and we do a thorough testing in Workday / Lightdash organizations before announcing it to the public.

cc: @TuringLovesDeathMetal

TuringLovesDeathMetal commented:
@IrakliJani I'm not sure what you mean 🤔 are you saying that in the `view` mode of a chart, you shouldn't be able to expand/collapse the groups?

IrakliJani commented:
@TuringLovesDeathMetal 

Apologies if the video was confusing...
Watch closely starting at 0:30 when I collapse/open subgroups - you'll notice the rows get duplicated.

I've been testing tables in various places, and one of the pages I examined is the "minimal explore view" page (the one where we take a screenshot for scheduled deliveries). For what it's worth, if this bug only exists on the minimal page table, then it's not a major issue because that page's interactivity is not crucial (it's intended for screenshots only).

IrakliJani commented:
just pushed one more commit which fixes the build 🙌

agha4to commented:
@IrakliJani, I did not know how to access `minimal` mode until now 🐤. Fixed this issue by using a unique row index for this type of table.

I searched to see if `getRowModel()` might be an issue anywhere else and found when "Show subtotals" and "Show number of results" are both checked, and there is only one page of data, the result total was incorrect. Now fixed.
![image](https://github.com/lightdash/lightdash/assets/138995798/3da15be2-4283-462e-b953-59998ae28d25)

@TuringLovesDeathMetal, changed "Show subtotals" checkbox behavior per above discussion.

IrakliJani commented:
@agha4to thanks for fixing that bug!

I just added few more commits and finished reviewing:
- improved types in random places
- fixed aggregation function bug

before:
![CleanShot 2024-03-29 at 17 16 41@2x](https://github.com/lightdash/lightdash/assets/962095/c4bca330-ce14-4478-8337-e8d00ddfebb5)
after:
![CleanShot 2024-03-29 at 17 16 10@2x](https://github.com/lightdash/lightdash/assets/962095/5e665e93-db80-40a1-aee2-5cb38912609f)

IrakliJani commented:
let me know if my changes make sense to you 🙌 
im going to approve the PR once the e2e pass

agha4to commented:
Lgtm. The before/after labels are switched above btw.
I introduced that subtotal bug when I switched to `some`, which I almost never use 😞 Nice find! I'm always learning to be more careful.

IrakliJani commented:
@agha4to cool!

Regarding "non-number sum": I believe it is acceptable to ignore any non-integer values and simply sum-up the remaining ones, as we are aggregating metric values that are either `COUNT` or `SUM`, which should always be numeric (almost always).

Regarding end-to-end testing: There was an issue with the rendering environment. I just reset the database, and now the end-to-end tests are all passing!

IrakliJani reviewed (approved):
amazing work @agha4to 🙌

IrakliJani commented:
@all-contributors add @agha4to for code
